File: leetcode/55.jump-game.py
# @leet start
class Solution:
    def canJump(self, nums: list[int]) -> bool:
        # Idea here: for any zeros we encounter, we need to make sure we
        # can jump over it. Can do this in one pass with constant
        # memory.
        current_max_jump = 0

        for num in nums:
            if num == 0:
                # Handle 0 here
                if current_max_jump < 2:
                    return False

            # Update current maximum jump distance
            current_max_jump = max(current_max_jump - 1, num)

        return True


# @leet end

File: leetcode/55.jump-game.js
// @leet start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function (nums) {
  const dp = Array(nums.length).fill(false);
  dp[nums.length - 1] = true;

  for (let i = nums.length - 2; i >= 0; i--) {
    for (let j = 1; j <= nums[i]; j++) {
      if (dp[i + j]) {
        dp[i] = true;
        break;
      }
    }
  }

  return dp[0];
};
// @leet end
